Assessment score mapping

Themba-Gqaza opened the pull request:
**PR Summary: Add Batch Scoring and Reporting to Simulation Runner**

**What this PR does**

This PR introduces a robust and scalable batch scoring feature that processes simulation outputs. After a simulation completes, this new logic calculates detailed scores for all applicable assessments and augments the original output with a full scoring report, including total scores, percentages, and min/max ranges.

**Key Changes**

- **New Scoring Functions (`tasks.py`):**
  - `score_assessment_from_simulation`: The main orchestration function that takes a simulation output and an assessment_id to produce a final score report.
  - `_score_and_format_turn`: A helper function that processes a single conversational turn, adding score and error data directly to it.
  - `_calculate_assessment_score_range`: A utility that dynamically calculates the minimum and maximum possible scores for any given assessment.

- **Updated Simulation Runner (`main`):**
  - The main function now orchestrates the scoring process after the simulation finishes.
  - It uses a scalable mapping (`scorable_assessments_map`) to identify and score multiple different assessment types (e.g., dma_assessment, knowledge-assessment).
  - The final output is a single, augmented JSON object containing the original simulation data enriched with the scoring results.

- **Comprehensive Pytest Suite (`test_tasks.py`):**
  - Adds a full test suite for the new scoring logic.
  - Uses `pytest.mark.parametrize` for concise and readable testing of various success and failure scenarios.
  - Validates the new, detailed scoring output, including `user_total_score`, percentages, and min/max ranges.

This feature decouples the simulation from the analysis, allowing for flexible and robust post-processing of assessment results.


Ouput example [need to fix the mapping part]: 
[run_output_with_scores.json](https://github.com/user-attachments/files/20709786/run_output_with_scores.json)

rudigiesler commented on src/ai4gd_momconnect_haystack/main.py:
```diff
@@ -52,16 +58,48 @@ def run_simulation():
+        diff_keys = [
+            k for k in user_context if user_context[k] != previous_context.get(k)
+        ]
+        if len(diff_keys) == 1:
```
What if it's more than 1? If the user provides us with 2 pieces of information in a single message

rudigiesler commented on src/ai4gd_momconnect_haystack/tasks.py:
```diff
@@ -158,3 +158,141 @@ def validate_assessment_answer(
+    turn["score"] = None
+    turn["score_error"] = "An unknown error occurred during scoring."
+
+    q_num_raw = turn.get("question_name")
```
The naming here in the `turn` is confusing, since this is not a question name, but a question number

rudigiesler commented on src/ai4gd_momconnect_haystack/main.py:
```diff
@@ -52,16 +58,48 @@ def run_simulation():
+                {
+                    "question_name": updated_field,
+                    "llm_utterance": contextualized_question,
+                    "user_utterance": user_response,
```
I can't find where we're using the `llm_utterance` and the `user_utterance` anywhere. This could maybe be changed to a simple map of answers, e.g.:
```
{
  0: "I disagree",
  1: "I agree",
}
```

Or we could rather have these be instances of dataclasses or a pydantic class, so that we know what fields to expect on both sides, and don't have to have a bunch of checks to assert the format of the unstructured dictionary.

Themba-Gqaza commented:
@rudigiesler I've addressed your comments. I've also updated the evaluation script to align with the latest simulation run output.

1. At the moment, running the evaluation script is still a manual process:
2. Run the simulation and respond using the gt_data.
3. Copy the run output to evaluations/ll_run_results.json.
4. Update the scenario_id in evaluations/data/ground_truth.json.
5. Run the evaluation script with: `docker-compose run --remove-orphans --rm python-app uv run python src/ai4gd_momconnect_haystack/evaluation/evaluator.py`

This workflow will be automated in a separate PR.

rudigiesler commented on src/ai4gd_momconnect_haystack/main.py:
```diff
@@ -95,13 +177,111 @@ def run_simulation():
+    # 2. Validate the IN-MEMORY raw output directly using Pydantic.
+    try:
+        simulation_output = [
+            AssessmentRun.model_validate(run) for run in raw_simulation_output
```
We could instead append `AssessmentRun` objects to the list (i.e. `raw_simulation_output` is of type `list[AssessmentRun]` instead of `list[dict]`), instead of appending dictionaries and then converting them to `AssessmentRun`s, but either way works.
